crates/red_knot_python_semantic/src/location.rs

```diff
@@ -0,0 +1,506 @@
+use crate::{
+    search::locate_name_on_type,
+    semantic_index::{definition::Definition, use_def_map, SemanticIndex},
+    Db, HasTy, SemanticModel,
+};
+use lsp_types::{Position, Range};
+use ruff_db::{
+    files::File,
+    source::{line_index, source_text},
+};
+use ruff_text_size::TextRange;
+use url::Url;
+
+// XXX should I just use an alias here? Not getting much value out of this huge import
+use ruff_python_ast::{
+    Arguments, Comprehension, Decorator, DictItem, ElifElseClause, ExceptHandler, Expr,
+    ExprAttribute, ExprAwait, ExprBinOp, ExprBoolOp, ExprCall, ExprCompare, ExprDict, ExprDictComp,
+    ExprFString, ExprGenerator, ExprIf, ExprLambda, ExprList, ExprListComp, ExprName, ExprNamed,
+    ExprSet, ExprSetComp, ExprSlice, ExprStarred, ExprSubscript, ExprTuple, ExprUnaryOp, ExprYield,
+    ExprYieldFrom, ExpressionRef, FString, FStringExpressionElement, FStringPart, FStringValue,
+    Identifier, Keyword, MatchCase, ModModule, Parameter, ParameterWithDefault, Parameters, Stmt,
+    StmtAnnAssign, StmtAssert, StmtAssign, StmtAugAssign, StmtClassDef, StmtDelete, StmtExpr,
+    StmtFor, StmtFunctionDef, StmtGlobal, StmtIf, StmtImport, StmtImportFrom, StmtMatch,
+    StmtNonlocal, StmtRaise, StmtReturn, StmtTry, StmtTypeAlias, StmtWhile, StmtWith, TypeParam,
+    TypeParamParamSpec, TypeParamTypeVar, TypeParamTypeVarTuple, TypeParams, WithItem,
+};
+
+#[derive(Debug)]
+pub enum DefLocation {
+    // XXX not including lsp_types in here would be good
+    // same for Url
+    Location { url: Url, range: Range },
+    Todo { s: String },
+}
+
+impl DefLocation {
+    pub(crate) fn from_definition<'db>(
+        definition: Definition<'db>,
+        index: &SemanticIndex<'db>,
+        db: &dyn Db,
+        file: File,
+    ) -> DefLocation {
+        let range = index.definition_range(&definition);
+        let final_range = ruff_range_to_lsp_range(db, file, range);
+        return DefLocation::Location {
+            url: definition.file(db).try_url(db.upcast()),
+            range: final_range,
+        };
+    }
+}
+
+fn ruff_range_to_lsp_range(db: &dyn Db, file: File, range: TextRange) -> lsp_types::Range {
+    let li = line_index(db.upcast(), file);
+    let contents = source_text(db.upcast(), file);
+    let loc_start = li.source_location(range.start(), &contents);
+    let loc_end = li.source_location(range.end(), &contents);
+    lsp_types::Range {
+        start: lsp_types::Position::new(
+            // XXX very wrong probably
+            loc_start.row.to_zero_indexed() as u32,
+            loc_start.column.to_zero_indexed() as u32,
+        ),
+        end: lsp_types::Position::new(
+            // XXX very wrong probably
+            loc_end.row.to_zero_indexed() as u32,
+            loc_end.column.to_zero_indexed() as u32,
+        ),
+    }
+}
+
+// this is a position as number of characters from the start
+// XXX find something to use instead of this
+pub struct CPosition(pub u64);
```

Comment:
I believe I could get away with using `TextRange` or something similar. At the very least, renaming it `CursorPosition` would make it clear-ish what I'm using it for